src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Delete.cs

```diff
@@ -0,0 +1,101 @@
+            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            {
+                watcher.IncludeSubdirectories = true;
+                watcher.NotifyFilter = NotifyFilters.DirectoryName;
+
+                string dirName = Path.Combine(nestedDir.Path, "dir");
+                Action action = () => Directory.Delete(dirName);
+                Action cleanup = () => Directory.CreateDirectory(dirName);
+                cleanup();
+
+                ExpectEvent(watcher, WatcherChangeTypes.Deleted, action, cleanup, dirName);
+            }
+        }
+
+        [Fact]
+        public void FileSystemWatcher_Directory_Delete_DeepDirectoryStructure()
+        {
+            // List of created directories
+            List<TempDirectory> lst = new List<TempDirectory>();
+
+            using (var testDirectory = new TempDirectory(GetTestFilePath()))
+            using (var dir = new TempDirectory(Path.Combine(testDirectory.Path, "dir")))
+            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            {
+                watcher.IncludeSubdirectories = true;
+                watcher.NotifyFilter = NotifyFilters.DirectoryName;
+
+                // Priming directory
+                lst.Add(new TempDirectory(Path.Combine(dir.Path, "dir")));
+
+                // Create a deep directory structure and expect things to work
+                for (int i = 1; i < 20; i++)
+                {
+                    // Test that the creation triggers an event correctly
+                    string dirPath = Path.Combine(lst[i - 1].Path, String.Format("dir{0}", i));
+                    Action action = () => Directory.Delete(dirPath);
+                    Action cleanup = () => Directory.CreateDirectory(dirPath);
+                    cleanup();
+
+                    ExpectEvent(watcher, WatcherChangeTypes.Deleted, action, cleanup);
+
+                    // Create the directory so subdirectories may be created from it.
+                    lst.Add(new TempDirectory(dirPath));
+                }
+            }
+        }
+
+        [ConditionalFact(nameof(CanCreateSymbolicLinks))]
+        public void FileSystemWatcher_Directory_Delete_SymLink()
+        {
+            using (var testDirectory = new TempDirectory(GetTestFilePath()))
+            using (var dir = new TempDirectory(Path.Combine(testDirectory.Path, "dir")))
+            using (var tempDir = new TempDirectory(GetTestFilePath()))
+            using (var watcher = new FileSystemWatcher(Path.GetFullPath(dir.Path), "*"))
+            {
+                // Make the symlink in our path (to the temp folder) and make sure an event is raised
+                string symLinkPath = Path.Combine(dir.Path, "link");
+                Action action = () => Directory.Delete(symLinkPath);
+                Action cleanup = () => Assert.True(CreateSymLink(tempDir.Path, symLinkPath, true));
+                cleanup();
```

Comment:
It's not a great name. I started with the tests that created a file/directory so ```cleanup``` made sense as a name that would then delete that file/directory.

For Deletion tests it makes less sense since it's the opposite of cleanup (initialization).

I couldn't think of a better name for a function that acts as the undoing of the action. I guess I could just rename it to be ```undo```.